### ISSUE TITLE
Nearby search with filter "Premium = yes" gives empty list (at least for base members)

### ISSUE THREAD

**Issue by Matze2** [cgeo/cgeo#11488]
### Describe your problem!

With the introduction of the new filter and/or the new nearby search it is not possible to see PMO caches in the nearby list at least for base members.

Even for base members the information that a PMO cache is nearby might be interesting when caching together with premium members and to be able to log from there.

### How to reproduce?

Log in as base member.
Go to Nearby list, open the filter and select "Yes" for Premium caches.
List is empty even if PMO caches are around.

### Actual result after these steps?

List is empty.

### Expected result after these steps?

List should contain nearby PMO caches if Premium filter is not set to "No".

### Reproducible

Yes

### c:geo Version

2021.08.17-NB-240646f

**Comment by Lineflyer** [comment]
I fear, that the new search we are now using is simply no longer returning PMO caches to BM...but I might be wrong.

**Comment by Matze2** [comment]
Not sure how c:geo gets the data but the nearby search on the web page still does it.

**Comment by eddiemuc** [comment]
I tried with current release (not nightly) and a basic user, it seems to work there. No clue currently what might be different in the nightly.

Nearby search now uses the same API as the live map search. Was this behavior observed there in the past?

**Comment by Matze2** [comment]
In the past it was anyway completely different (with paging), but it seems the "Nearby" content is still the same, e.g. here
https://www.geocaching.com/play/search?lat=48&lng=11&origin=48,11&utr=false

The only difference I saw that it is now not possible anymore to exclude already found caches as basic member on the web page.

**Comment by eddiemuc** [comment]
Have to correct myself. It seems c:geo somehow uses old credentials in background for cache search when switching from premium to basic user until c:geo is restarted. I can confirm the problem.

Need to check in code as soon as I can get my hands on a dev machine...

**Comment by Matze2** [comment]
Do we use this URL for Nearby?
https://www.geocaching.com/api/proxy/web/search/v2?box=48.086336%2C10.968132%2C48.032183%2C11.297722&skip=0&take=500&asc=true&sort=distance
This includes PMO entries (premiumOnly: true). So maybe it is only a bug.

**Comment by eddiemuc** [comment]
> I fear, that the new search we are now using is simply no longer returning PMO caches to BM...but I might be wrong.

I fear you are right. As basic member I also don't see any premium caches in the map on the web page (and this makes sense since basic members don't get any coordinates for premium caches).

I fear the only way to get this feature back would be to reactivate the old "web page scraping" search - with all restrictions coming with this. Damn!

**Comment by Matze2** [comment]
See my post above. The JSON response contains PMO entries. But of course without coordinates. So the information is provided but not shown on the map.

**Comment by Matze2** [comment]
Also here PMO entries are shown in the list on the left.
https://www.geocaching.com/play/map?lat=48.0000&lng=11.0000&zoom=14&asc=true&sort=distance&clean=true

**Comment by eddiemuc** [comment]
> See my post above. The JSON response contains PMO entries. But of course without coordinates. So the information is provided but not shown on the map.

You're right. And yes, this is the url used. Probably caches without any coordinate are currently sorted out though (in previous releases, this Url was NOT used for nearby search but only for map viewer). 

And yes, since the info is there it can be incorporated. Hooray, no return to scraping necessary!

**Comment by Matze2** [comment]
No this paging-less variant is much better.

**Comment by Matze2** [comment]
By the way. How big is the "box" (distance, number of caches) that is used for the search and is it configurable?

**Comment by eddiemuc** [comment]
Ok, I searched in the code via GitHub App one mobile phone. It seems that all caches without a coordinate are filtered out very early when results are received in method `GcWebApi.searchCaches` (around line 764).

It might be enough to just allow caches without coordinates to pass here to fix the problem. Of course this must be tested carefully for any side effects.

**Comment by Lineflyer** [comment]
> By the way. How big is the "box" (distance, number of caches) that is used for the search and is it configurable?

That is a relevant question, and recently asked by users.
Somehow users got used to the pagination (20 packs) as a convenient feature to just store those to have the closest, while now it requires some manual selection or filtering to achieve this (or store all 200 caches).
Maybe we could use a clever boxing or alike, will transfer in a new issue.

**Comment by eddiemuc** [comment]
> By the way. How big is the "box" (distance, number of caches) that is used for the search and is it configurable?

For nearby search, only the origin is set (no max distance), as well as a limit of 200 caches to return per connector. So depending on the connectors handling of server requests, if you have 3 active connectors you should end up with 600 caches - minus filters which you might apply in addition to the ones supported by the connector, and minus whatever logic the connectors server uses to determine the result count it returns.

The 200 is currently hard-coded.

**Comment by Matze2** [comment]
> It might be enough to just allow caches without coordinates to pass here to fix the problem. Of course this must be tested carefully for any side effects.

I would volunteer to test at least my use cases.

**Comment by eddiemuc** [comment]
In PR #11496 I made the GC Web API also pass through caches without coordinates, and it seems to work fine.

However, there is another problem: the Web API does also not deliver estimated distance and direction for premium caches any more (while the web search does)...

I implemented a heuristic to get distances out of the list (exploiting the fact that the returned list is sorted by distance, I assign the middle distance of two basic caches "surrounding" a premium cache in the list.)

However I could think of no way to implement a heuristic for the direction of the cache (N/S/W/E/...) without having its coordinates. Also, the heuristic to calculate distance might be good enough for sorting, but it might be wrong of course.

Also in the map view on gc.com, premium caches ARE indeed shown in the list on the left but without distance/direction information (https://www.geocaching.com/play/map?lat=48.0000&lng=11.0000&zoom=14&asc=true&sort=distance&clean=true). The old web-parsing-search still has this information.

Don't know exactly how to continue from here. I vote for merging the PR (after review of course) so basic users have premium caches back. Still this is a functional degradation - which I currently have no idea how to solve without reverting online search back to the legacy version (and then without online filter possibilities of course)

**Comment by Matze2** [comment]
Many thanks for your work, Eddie.

In my opinion there are many disadvantages with the legacy version
- the parsing of the JSON response should be much more stable than the web version
- it's more resource intensive
- it might be removed at any time

Therefore I don't think that there is really a choice in the long run.
The distance and direction information are only not that crucial in my opinion (direction less than distance). The heuristic you implemented is a nice fallback.

In short: I could live with the solution.

**Comment by Lineflyer** [comment]
> Also in the map view on gc.com, premium caches ARE indeed shown in the list on the left but without distance/direction information (https://www.geocaching.com/play/map?lat=48.0000&lng=11.0000&zoom=14&asc=true&sort=distance&clean=true). The old web-parsing-search still has this information.

IMHO the map view for basic members in c:geo never showed premium caches on the map. Those PM caches have ever since only been listed in searches (the initial topic) and when directly using its geocode for a search (but failing to load the details then of course).

**Comment by eddiemuc** [comment]
> IMHO the map view for basic members in c:geo never showed premium caches on the map. Those PM caches have ever since only been listed in searches (the initial topic) and when directly using its geocode for a search (but failing to load the details then of course).

@Lineflyer that's true, I am only talking about the list view. But the old search showed information about distance and direction of premium caches in the list view for basic members (as parsed from the search result website) while the new search doesn't have this information (it is not provided via the webApi). Especially distance is also relevant for list sorting since it is the default/most-used sorting method.

For distance I implemented a new heuristic (as described above), for direction I don't think a heuristic is possible.

**Comment by Matze2** [comment]
Would you mind to merge that commit also to master? I am using the nightlies of c:geo.

**Comment by Lineflyer** [comment]
it seems conflict resolution is required for that, so someone needs to look into the merge this time.

**Comment by eddiemuc** [comment]
If conflicts are pasted here I might be able to remote-assist their resolution. But currently no chance to do it myself (no PC, no stable internet)

**Comment by rsudev** [comment]
I can take a look.

**Comment by moving-bits** [comment]
Thanks, @rsudev.
I won't be back at my dev machine before Sunday evening.

**Comment by eddiemuc** [comment]
Thanks a lot @rsudev for helping out!

**Comment by Matze2** [comment]
Awesome job, guys! Thank you very much for the quick fix!
